File: extension/src/score/verificationConcentration.ts
export interface ReviewForVerification {
  rating: number | null;
  verified: boolean | null;
  insideBurst: boolean;
}

export interface VerificationConcentrationResult {
  lift: number | null;
  baseCount: number;
}

export function verificationConcentration(
  reviews: readonly ReviewForVerification[],
): VerificationConcentrationResult {
  const known = reviews.filter((review) => review.verified !== null);
  if (known.length === 0) {
    return { lift: null, baseCount: 0 };
  }

  const overallUnverifiedRate = known.filter((review) => review.verified === false).length /
    known.length;

  const stratum = known.filter((review) => review.rating === 5 && review.insideBurst);
  const baseCount = stratum.length;
  if (baseCount === 0 || overallUnverifiedRate === 0) {
    return { lift: null, baseCount };
  }

  const stratumUnverifiedRate = stratum.filter((review) => review.verified === false).length /
    baseCount;

  return { lift: stratumUnverifiedRate / overallUnverifiedRate, baseCount };
}

File: research/verdict_research/features/verification_concentration.py
from dataclasses import dataclass


@dataclass
class ReviewForVerification:
    rating: int | None
    verified: bool | None
    inside_burst: bool


@dataclass
class VerificationConcentrationResult:
    lift: float | None
    base_count: int


def verification_concentration(
    reviews: list[ReviewForVerification],
) -> VerificationConcentrationResult:
    known = [review for review in reviews if review.verified is not None]
    if not known:
        return VerificationConcentrationResult(lift=None, base_count=0)

    overall_unverified_rate = sum(1 for review in known if review.verified is False) / len(known)

    stratum = [review for review in known if review.rating == 5 and review.inside_burst]
    base_count = len(stratum)
    if base_count == 0 or overall_unverified_rate == 0:
        return VerificationConcentrationResult(lift=None, base_count=base_count)

    stratum_unverified_rate = sum(1 for review in stratum if review.verified is False) / base_count

    return VerificationConcentrationResult(
        lift=stratum_unverified_rate / overall_unverified_rate,
        base_count=base_count,
    )
